src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java

```diff
@@ -486,41 +486,63 @@ private void handleCurrentStateBroken(Optional<State> desiredState, State previo
             return;
         }
 
-        long currentStateGeneration = stateGeneration.incrementAndGet();
-        replaceBackingTask(() -> {
-            if (!State.NEW.equals(getState()) || getStateGeneration().get() != currentStateGeneration) {
-                // Bail out if we're not in the expected state
+        if (desiredState.get().equals(State.RUNNING)) {
+            // if there is already a install() task running, do nothing.
+            Future<?> currentTask = backingTask.get();
+            if (currentTask != null && !currentTask.isDone()) {
                 return;
             }
-            try {
-                greengrassService.install();
-            } catch (InterruptedException t) {
-                logger.atWarn("service-install-interrupted").log("Service interrupted while running install");
-            } catch (Throwable t) {
-                greengrassService.serviceErrored(t);
-            }
-        }, LIFECYCLE_INSTALL_NAMESPACE_TOPIC);
 
-        Integer installTimeOut = getTimeoutConfigValue(
-                LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+            long currentStateGeneration = stateGeneration.incrementAndGet();
 
-        try {
-            backingTask.get().get(installTimeOut, TimeUnit.SECONDS);
-            if (!State.ERRORED.equals(lastReportedState.get())) {
-                internalReportState(State.INSTALLED);
-            }
-        } catch (ExecutionException ee) {
-            greengrassService.serviceErrored(ee);
-        } catch (TimeoutException te) {
-            greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
-        } finally {
-            stopBackingTask();
+            Integer timeout = getTimeoutConfigValue(
+                    LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+            Future<?> schedule =
+                    greengrassService.getContext().get(ScheduledExecutorService.class).schedule(() -> {
+                        if (getState().equals(State.NEW) && currentStateGeneration == getStateGeneration().get()) {
+                            greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
+                        }
+                    }, timeout, TimeUnit.SECONDS);
+
+            replaceBackingTask(() -> {
+                if (!State.NEW.equals(getState()) || getStateGeneration().get() != currentStateGeneration) {
+                    // Bail out if we're not in the expected state
+                    return;
+                }
+                try {
+                    greengrassService.install();
+                    if (!State.ERRORED.equals(lastReportedState.get()) && !schedule.isDone()) {
+                        internalReportState(State.INSTALLED);
+                        schedule.cancel(true);
+                    }
+                } catch (InterruptedException t) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)